### ISSUE TITLE
Document the pros and cons between using rootfs vs initrd

### ISSUE THREAD

**Issue by raravena80** [kata-containers/documentation#167]
Not sure if this is documented, but opening to keep track. Here are some items to consider (from the discussion on slack)

- On a host running multiple Kata VMs the disk image will be shared in the host page cache.
- When using initramfs the memory is duplicated until Kernel Samepage Merging decides to deduplicate it.
- There could be memory savings if you using a disk image, depending on the type of containers.

**Comment by grahamwhaley** [comment]
Thanks for opening @raravena80 
There is a little bit of historical info on DAX over at https://github.com/clearcontainers/runtime/blob/master/docs/architecture/architecture.md#dax that might help.
I believe also that, generally, the initrd images can be smaller as (in their present invocation anyhow) I think they run the agent directly as the init (pid1) process, whereas the image files include systemd to initialise the system. I don't think this is necessarily a hard requirement for either method (image or initrd vs systemd or agent==init), I think that is just how we do it right now, that is image+systemd and initrd+init==agent.
I think maybe @bergwolf can add more details about some of the advantages of initrd.

**Comment by bergwolf** [comment]
Yes, I can add a few bits about disk images vs. initrd (or DAX+KSM vs. vm templating since they are mostly used together and @raravena80 already outlined DAX+KSM):
- The size of initrd is much (x10 times) smaller and is thus easier to ship.
- DAX requires kernel support that means it does not work on old guest kernels.
- VM templating requires QEMU support that means it does not work for old QEMUs. That said, I think we should patch qemu-lite to support it at least.
- Disk images rely on KSM to share guest kernel image pages and agent memory, and DAX to share disk image pages. While initrd makes use of vm templating, which shares the initial memory of all guests, including kernel, initrd and agent memory, and thus achieves similar effects like KSM+DAX.
- The initial memory saving is close between initrd+vm templating and disk image+KSM.
- KSM costs host CPU resource and there is no runtime cost with vm templating.
- When KSM is running, it also pollutes and invalidates CPU cache lines, which can be bad for guest performance.

In summary, it is all about trade-offs:
- If you just want memory saving and no matter what, disk image + KSM is your preference.
- If you do not want KSM to interfere with your guest performance and still want to save memory, initrd+vm templating is the way to go.
- If you do not care much about memory saving nor the size of image to install, choose disk images w/o KSM.
- If you do not care much about memory saving but want easy shipment, choose initrd w/o vm templating.
